File: src/zig/memory/allocator.zig
//! Custom Memory Allocator for SigmaOS
//! Implements memory allocation without relying on std
//! Uses buddy system algorithm for efficient memory management

const BLOCK_SIZE: usize = 4096;
const MAX_ORDER: usize = 10;

/// Memory block structure
pub const MemoryBlock = struct {
    size: usize,
    used: bool,
    next: ?*MemoryBlock,
    prev: ?*MemoryBlock,
};

/// Buddy system memory allocator
pub const BuddyAllocator = struct {
    free_lists: [?*MemoryBlock][MAX_ORDER + 1],
    total_memory: usize,
    used_memory: usize,
    heap_start: [*]u8,
    heap_size: usize,
};

pub fn initAllocator(allocator: *BuddyAllocator, heap_start: [*]u8, heap_size: usize) void {
    allocator.free_lists = [_]null;
    allocator.total_memory = heap_size;
    allocator.used_memory = 0;
    allocator.heap_start = heap_start;
    allocator.heap_size = heap_size;

    // Initialize free lists
    var remaining = heap_size;
    var current = heap_start;

    while (remaining >= BLOCK_SIZE) {
        var order = calculateOrder(remaining);
        var block_size = BLOCK_SIZE * (1 << order);

        var block = @ptrCast(*MemoryBlock, current);
        block.size = block_size;
        block.used = false;
        block.next = null;
        block.prev = null;

        addToFreeList(allocator, block, order);

        current = current + block_size;
        remaining -= block_size;
    }
}

fn calculateOrder(size: usize) usize {
    var order: usize = 0;
    var block_size = BLOCK_SIZE;
    while (block_size < size and order < MAX_ORDER) {
        block_size *= 2;
        order += 1;
    }
    return order;
}

fn addToFreeList(allocator: *BuddyAllocator, block: *MemoryBlock, order: usize) void {
    block.next = allocator.free_lists[order];
    if (allocator.free_lists[order]) |head| {
        head.prev = block;
    }
    allocator.free_lists[order] = block;
}

fn removeFromFreeList(allocator: *BuddyAllocator, block: *MemoryBlock, order: usize) void {
    const prev = block.prev;
    const next = block.next;

    if (prev) |prev_block| {
        prev_block.next = next;
    } else {
        allocator.free_lists[order] = next;
    }

    if (next) |next_block| {
        next_block.prev = prev;
    }

    block.prev = null;
    block.next = null;
}

pub fn allocate(allocator: *BuddyAllocator, size: usize) ?[*]u8 {
    const aligned_size = (size + @sizeOf(MemoryBlock) + 15) & ~@as(usize, 15);
    const order = calculateOrder(aligned_size);

    // Find a free block of sufficient size
    var current_order = order;
    var block: ?*MemoryBlock = null;

    while (current_order <= MAX_ORDER) {
        if (allocator.free_lists[current_order]) |free_block| {
            block = free_block;
            break;
        }
        current_order += 1;
    }

    if (block == null) {
        return null;
    }

    removeFromFreeList(allocator, block, current_order);

    // Split block if necessary
    while (current_order > order) {
        current_order -= 1;
        const split_size = BLOCK_SIZE * (1 << current_order);
        const buddy = @ptrCast(*MemoryBlock, @intToPtr([*]u8, block) + split_size);

        buddy.size = split_size;
        buddy.used = false;
        buddy.next = null;
        buddy.prev = null;

        addToFreeList(allocator, buddy, current_order);

        block.size = split_size;
    }

    block.used = true;
    allocator.used_memory += block.size;

    return @intToPtr([*]u8, block) + @sizeOf(MemoryBlock);
}

pub fn deallocate(allocator: *BuddyAllocator, ptr: [*]u8) void {
    const block = @ptrCast(*MemoryBlock, ptr - @sizeOf(MemoryBlock));

    if (!block.used) {
        return; // Already freed
    }

    block.used = false;
    allocator.used_memory -= block.size;

    const size = block.size;
    const order = calculateOrder(size);

    // Try to merge with buddy
    var current_block = block;
    var current_order = order;

    while (current_order < MAX_ORDER) {
        const buddy = getBuddy(current_block, current_order);

        if (buddy.used or buddy.size != current_block.size) {
            break;
        }

        // Remove buddy from free list
        removeFromFreeList(allocator, buddy, current_order);

        // Merge blocks
        if (current_block < buddy) {
            current_block.size *= 2;
        } else {
            buddy.size *= 2;
            current_block = buddy;
        }

        current_order += 1;
    }

    addToFreeList(allocator, current_block, current_order);
}

fn getBuddy(block: *MemoryBlock, order: usize) *MemoryBlock {
    const block_addr = @intFromPtr(block);
    const size = BLOCK_SIZE * (1 << order);
    const buddy_addr = block_addr ^ size;
    return @ptrFromInt(MemoryBlock, buddy_addr);
}

pub fn getStats(allocator: *BuddyAllocator) MemoryStats {
    return MemoryStats{
        .total = allocator.total_memory,
        .used = allocator.used_memory,
        .free = allocator.total_memory - allocator.used_memory,
    };
}

pub const MemoryStats = struct {
    total: usize,
    used: usize,
    free: usize,
};

// External allocator functions (would be provided by kernel)
extern fn kernelAlloc(size: usize) [*]u8;
extern fn kernelFree(ptr: [*]u8);

File: src/nim/memory/allocator.nim
# Custom Memory Allocator for SigmaOS
# Implements memory allocation without relying on std
# Uses buddy system algorithm for efficient memory management

const BLOCK_SIZE = 4096
const MAX_ORDER = 10

type
  MemoryBlock = object
    size: int
    used: bool
    next: ptr MemoryBlock
    prev: ptr MemoryBlock

  BuddyAllocator = object
    freeLists: array[MAX_ORDER + 1, ptr MemoryBlock]
    totalMemory: int
    usedMemory: int
    heapStart: pointer
    heapSize: int

proc initAllocator(allocator: ptr BuddyAllocator, heapStart: pointer, heapSize: int) =
  allocator.freeLists = cast[array[MAX_ORDER + 1, ptr MemoryBlock]]([nil, nil, nil, nil, nil, nil, nil, nil, nil, nil, nil])
  allocator.totalMemory = heapSize
  allocator.usedMemory = 0
  allocator.heapStart = heapStart
  allocator.heapSize = heapSize
  
  var remaining = heapSize
  var current = heapStart
  
  while remaining >= BLOCK_SIZE:
    var order = calculateOrder(remaining)
    var blockSize = BLOCK_SIZE * (1 shl order)
    
    var block = cast[ptr MemoryBlock](current)
    block.size = blockSize
    block.used = false
    block.next = nil
    block.prev = nil
    
    addToFreeList(allocator, block, order)
    
    current = cast[pointer](cast[int](current) + blockSize)
    remaining -= blockSize

proc calculateOrder(size: int): int =
  var order = 0
  var blockSize = BLOCK_SIZE
  while blockSize < size and order < MAX_ORDER:
    blockSize = blockSize * 2
    order += 1
  return order

proc addToFreeList(allocator: ptr BuddyAllocator, block: ptr MemoryBlock, order: int) =
  block.next = allocator.freeLists[order]
  if not isNil(allocator.freeLists[order]):
    allocator.freeLists[order].prev = block
  allocator.freeLists[order] = block

proc removeFromFreeList(allocator: ptr BuddyAllocator, block: ptr MemoryBlock, order: int) =
  let prev = block.prev
  let next = block.next
  
  if not isNil(prev):
    prev.next = next
  else:
    allocator.freeLists[order] = next
  
  if not isNil(next):
    next.prev = prev
  
  block.prev = nil
  block.next = nil

proc allocate(allocator: ptr BuddyAllocator, size: int): pointer =
  let alignedSize = (size + sizeof(MemoryBlock) + 15) and (not 15)
  let order = calculateOrder(alignedSize)
  
  var currentOrder = order
  var block: ptr MemoryBlock = nil
  
  while currentOrder <= MAX_ORDER:
    if not isNil(allocator.freeLists[currentOrder]):
      block = allocator.freeLists[currentOrder]
      break
    currentOrder += 1
  
  if isNil(block):
    return nil
  
  removeFromFreeList(allocator, block, currentOrder)
  
  while currentOrder > order:
    currentOrder -= 1
    let splitSize = BLOCK_SIZE * (1 shl currentOrder)
    let buddy = cast[ptr MemoryBlock](cast[int](block) + splitSize)
    
    buddy.size = splitSize
    buddy.used = false
    buddy.next = nil
    buddy.prev = nil
    
    addToFreeList(allocator, buddy, currentOrder)
    
    block.size = splitSize
  
  block.used = true
  allocator.usedMemory += block.size
  
  return cast[pointer](cast[int](block) + sizeof(MemoryBlock))

proc deallocate(allocator: ptr BuddyAllocator, ptr: pointer) =
  let block = cast[ptr MemoryBlock](cast[int](ptr) - sizeof(MemoryBlock))
  
  if not block.used:
    return
  
  block.used = false
  allocator.usedMemory -= block.size
  
  let size = block.size
  let order = calculateOrder(size)
  
  var currentBlock = block
  var currentOrder = order
  
  while currentOrder < MAX_ORDER:
    let buddy = getBuddy(currentBlock, currentOrder)
    
    if buddy.used or buddy.size != currentBlock.size:
      break
    
    removeFromFreeList(allocator, buddy, currentOrder)
    
    if cast[int](currentBlock) < cast[int](buddy):
      currentBlock.size = currentBlock.size * 2
    else:
      buddy.size = buddy.size * 2
      currentBlock = buddy
    
    currentOrder += 1
  
  addToFreeList(allocator, currentBlock, currentOrder)

proc getBuddy(block: ptr MemoryBlock, order: int): ptr MemoryBlock =
  let blockAddr = cast[int](block)
  let blockSize = BLOCK_SIZE * (1 shl order)
  let buddyAddr = blockAddr xor blockSize
  return cast[ptr MemoryBlock](buddyAddr)

proc getStats(allocator: ptr BuddyAllocator): tuple[total: int, used: int, free: int] =
  return (allocator.totalMemory, allocator.usedMemory, allocator.totalMemory - allocator.usedMemory)

# External allocator functions (would be provided by kernel)
proc kernelAlloc(size: int): pointer {.importc, header: "<kernel.h>".}
proc kernelFree(ptr: pointer) {.importc, header: "<kernel.h>".}
